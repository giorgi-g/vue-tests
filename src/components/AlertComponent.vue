<template>
  <div class="alert">
    <p>{{ title }}</p>
    <p>Time left {{ timeLeft }}</p>
  </div>
</template>

<script>
export default {
  props: {
    title: {
      type: String,
      required: false,
      default: "TIMER"
    }
  },
  data() {
    return {
      interval: undefined,
      counter: 0,
      timer: 5
    };
  },
  computed: {
    timeLeft() {
      return this.timer - this.counter;
    }
  },
  mounted() {
    this.interval = setInterval(() => {
      this.counter++;
      if (this.counter === this.timer) this.$destroy();
    }, 1000);
  },
  beforeDestroy() {
    clearInterval(this.interval);
    console.log("Alert removed!");
  },
  destroyed() {
    this.$el.remove();
  }
};
</script>
