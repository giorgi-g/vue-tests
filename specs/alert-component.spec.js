import AlertComponent from "@/components/AlertComponent";
import { mount } from "@vue/test-utils";

test("Mount an AlertComponent component", () => {
  const wrapper = mount(AlertComponent, {
    propsData: {
      title: "The Timer!"
    }
  });
  console.log("Wrapper", wrapper);
  /**
   * To update the snapshot
   * @run yarn jest -u
   * */
  expect(wrapper.html()).toMatchSnapshot();
});

jest.useFakeTimers();

describe("Lifecycle methods", () => {
  test("Mounted assigns interval", () => {
    const wrapper = mount(AlertComponent);
    // console.log("wrapper.vm.interval", wrapper.vm.interval);
    expect(wrapper.vm.interval).not.toBe(undefined);
  });

  test("counter workds", () => {
    const wrapper = mount(AlertComponent);
    expect(wrapper.vm.counter).toBe(0);
    // console.log("0. wrapper.vm.counter", wrapper.vm.counter);
    jest.advanceTimersByTime(1000);
    expect(wrapper.vm.counter).toBe(1);
    // console.log("1. wrapper.vm.counter", wrapper.vm.counter);
    jest.advanceTimersByTime(1000);
    expect(wrapper.vm.counter).toBe(2);
    // console.log("2. wrapper.vm.counter", wrapper.vm.counter);
  });

  test("instance gets destroyed", () => {
    const beforeDestroyedSpy = jest.spyOn(AlertComponent, "beforeDestroy");
    const wrapper = mount(AlertComponent);
    wrapper.vm.counter = wrapper.vm.timer - 1;
    jest.advanceTimersByTime(1000);
    expect(beforeDestroyedSpy).toHaveBeenCalled();
  });
});
