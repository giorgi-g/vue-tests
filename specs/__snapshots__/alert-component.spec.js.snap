// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mount an AlertComponent component 1`] = `
<div class="alert">
  <p>The Timer!</p>
  <p>Time left 5</p>
</div>
`;
